File: web/src/api/vpc.ts
/**
 * VPC 相关 API（交换机 / 安全组 / VM 绑定）
 * 对应后端 /api/vpc、/api/vm/:name/vpc 路由组
 */
import service from './client'
import type { ApiResponse } from '@/types/api'

// ==================== 类型定义 ====================

/** VPC 交换机 */
export interface VpcSwitch {
  id: number
  username: string
  is_system: boolean
  name: string
  bridge_name: string
  bridge_mode: string // nat / bridge
  bridge_vlan_id: number
  allow_promiscuous?: boolean
  allow_mac_change?: boolean
  allow_forged_transmits?: boolean
  vlan_id: number
  cidr: string
  gateway_ip: string
  dhcp_start: string
  dhcp_end: string
  traffic_down_gb: number
  traffic_up_gb: number
  bandwidth_mbps?: number // 兼容旧版单总带宽字段
  bandwidth_down_mbps: number
  bandwidth_up_mbps: number
  used_traffic_down?: number // 字节
  used_traffic_up?: number // 字节
  used_traffic_down_gb?: string
  used_traffic_up_gb?: string
  is_limited_down?: boolean
  is_limited_up?: boolean
  effective_bandwidth_down_mbps?: number
  effective_bandwidth_up_mbps?: number
}

/** 交换机创建/编辑请求 */
export interface VpcSwitchPayload {
  username?: string
  name: string
  bridge_name?: string
  bridge_vlan_id?: number
  allow_promiscuous?: boolean
  allow_mac_change?: boolean
  allow_forged_transmits?: boolean
  cidr?: string
  gateway_ip?: string
  dhcp_start?: string
  dhcp_end?: string
  traffic_down_gb?: number
  traffic_up_gb?: number
  bandwidth_mbps?: number
  bandwidth_down_mbps?: number
  bandwidth_up_mbps?: number
}

/** VPC 流量/带宽配额 */
export interface VpcQuota {
  username: string
  max_traffic_down: number
  max_traffic_up: number
  allocated_traffic_down: number
  allocated_traffic_up: number
  remaining_traffic_down: number
  remaining_traffic_up: number
  max_bandwidth_down: number
  max_bandwidth_up: number
  allocated_bandwidth_down: number
  allocated_bandwidth_up: number
  remaining_bandwidth_down: number
  remaining_bandwidth_up: number
}

/** 交换机下绑定的虚拟机 */
export interface VpcSwitchVM {
  vm_name: string
  username: string
  interface_order: number
}

/** 安全组规则 */
export interface VpcSecurityGroupRule {
  id: number
  security_group_id: number
  direction: string // ingress / egress
  protocol: string // tcp / udp / icmp / all
  port_start: number
  port_end: number
  target_type: string
  target_value: string
  remark: string
}

/** 安全组 */
export interface VpcSecurityGroup {
  id: number
  username: string
  vm_name?: string
  name: string
  is_default: boolean
  is_vm_scoped?: boolean
  remark?: string
  rules?: VpcSecurityGroupRule[]
}

/** VM 与交换机/安全组的绑定 */
export interface VpcVMBinding {
  id: number
  vm_name: string
  username: string
  switch_id: number
  security_group_id: number
  interface_order: number
  nic_model: string
  bandwidth_inbound_avg: number
  bandwidth_outbound_avg: number
}

/** 轻量云 VM 配额（来自绑定信息） */
export interface LightweightVMQuota {
  traffic_down_gb: number
  traffic_up_gb: number
  bandwidth_down_mbps: number
  bandwidth_up_mbps: number
  max_port_forwards: number
  max_snapshots: number
  max_runtime_hours: number
  used_runtime_seconds: number
  used_traffic_down: number
  used_traffic_up: number
  used_traffic_down_gb: string
  used_traffic_up_gb: string
  used_runtime_display: string
  remaining_runtime_seconds: number
  remaining_runtime_display: string
  runtime_quota_reached: boolean
  current_net_rx_rate: string
  current_net_tx_rate: string
  is_limited_down: boolean
  is_limited_up: boolean
  used_port_forwards: number
  used_snapshots: number
}

/** VM VPC 绑定信息（/vm/:name/vpc 响应） */
export interface VpcBindingInfo {
  owner_username?: string
  binding?: VpcVMBinding | null
  bindings?: VpcVMBinding[]
  switch?: VpcSwitch | null
  security_group?: VpcSecurityGroup | null
  groups?: VpcSecurityGroup[]
  switches?: VpcSwitch[]
  lightweight_quota?: LightweightVMQuota | null
}

// ==================== 接口 ====================

/** 获取 VPC 交换机列表 */
export function getVPCSwitches(params?: { username?: string; keyword?: string }) {
  return service.get<unknown, ApiResponse<VpcSwitch[]>>('/vpc/switches', { params, silent: true })
}

/** 获取 VPC 配额（剩余/已分配流量与带宽） */
export function getVPCQuota(params?: { username?: string }) {
  return service.get<unknown, ApiResponse<VpcQuota>>('/vpc/quota', { params, silent: true })
}

/** 创建交换机 */
export function createVPCSwitch(data: VpcSwitchPayload) {
  return service.post<unknown, ApiResponse<unknown>>('/vpc/switches', data)
}

/** 更新交换机 */
export function updateVPCSwitch(id: number, data: VpcSwitchPayload) {
  return service.put<unknown, ApiResponse<unknown>>(`/vpc/switches/${id}`, data)
}

/** 删除交换机（force=true 时强制移除绑定 VM 的网卡） */
export function deleteVPCSwitch(id: number, force = false) {
  return service.delete<unknown, ApiResponse<unknown>>(`/vpc/switches/${id}`, {
    params: { force: force ? 'true' : '' },
  })
}

/** 获取交换机下绑定的虚拟机 */
export function getVPCSwitchVMs(id: number) {
  return service.get<unknown, ApiResponse<VpcSwitchVM[]>>(`/vpc/switches/${id}/vms`)
}

/** 重置交换机本月流量计数 */
export function resetVPCSwitchTraffic(id: number) {
  return service.post<unknown, ApiResponse<unknown>>(`/vpc/switches/${id}/traffic/reset`)
}

/** 获取安全组列表 */
export function getVPCSecurityGroups(params?: { username?: string; keyword?: string }) {
  return service.get<unknown, ApiResponse<VpcSecurityGroup[]>>('/vpc/security-groups', {
    params,
    silent: true,
  })
}

/** 创建安全组 */
export function createVPCSecurityGroup(data: { username?: string; name: string; remark?: string }) {
  return service.post<unknown, ApiResponse<unknown>>('/vpc/security-groups', data)
}

/** 更新安全组 */
export function updateVPCSecurityGroup(
  id: number,
  data: { username?: string; name: string; remark?: string },
) {
  return service.put<unknown, ApiResponse<unknown>>(`/vpc/security-groups/${id}`, data)
}

/** 删除安全组 */
export function deleteVPCSecurityGroup(id: number) {
  return service.delete<unknown, ApiResponse<unknown>>(`/vpc/security-groups/${id}`)
}

/** 预览 VPC ACL（nftables 规则文本） */
export function previewVPCACL() {
  return service.get<unknown, ApiResponse<string>>('/vpc/acl/preview')
}

/** 应用 VPC ACL（重建防火墙规则，高风险） */
export function applyVPCACL() {
  return service.post<unknown, ApiResponse<unknown>>('/vpc/acl/apply')
}

/** 添加安全组规则 */
export function addVPCSecurityGroupRule(
  id: number,
  data: {
    direction: string
    protocol: string
    port_start: number
    port_end: number
    target_type: string
    target_value: string
    remark: string
  },
) {
  return service.post<unknown, ApiResponse<unknown>>(`/vpc/security-groups/${id}/rules`, data)
}

/** 删除安全组规则 */
export function deleteVPCSecurityGroupRule(id: number) {
  return service.delete<unknown, ApiResponse<unknown>>(`/vpc/security-groups/rules/${id}`)
}

/** 获取 VM 的 VPC 绑定信息 */
export function getVMVPCBinding(name: string) {
  return service.get<unknown, ApiResponse<VpcBindingInfo>>(
    `/vm/${encodeURIComponent(name)}/vpc`,
    { silent: true },
  )
}

/** 更新 VM 的 VPC 绑定（切换交换机/安全组） */
export function bindVMVPC(name: string, data: { switch_id: number; security_group_id: number }) {
  return service.put<unknown, ApiResponse<unknown>>(`/vm/${encodeURIComponent(name)}/vpc`, data)
}

/** 切换 VM 安全组 */
export function switchVMSecurityGroup(name: string, securityGroupID: number) {
  return service.put<unknown, ApiResponse<unknown>>(
    `/vm/${encodeURIComponent(name)}/security-group`,
    { security_group_id: securityGroupID },
  )
}

// ==================== 多网口管理（仅管理员） ====================

/** 网口信息（绑定 + 交换机 + 安全组） */
export interface VMInterfaceInfo {
  binding: VpcVMBinding
  switch?: VpcSwitch | null
  security_group?: VpcSecurityGroup | null
  mac?: string
}

/** 网口新增/更新请求 */
export interface VMInterfacePayload {
  switch_id: number
  security_group_id: number
  nic_model: string
  bandwidth_inbound_avg: number
  bandwidth_outbound_avg: number
}

/** 获取虚拟机网口列表（仅管理员） */
export function listVMInterfaces(name: string) {
  return service.get<unknown, ApiResponse<VMInterfaceInfo[]>>(
    `/vm/${encodeURIComponent(name)}/interfaces`,
    { silent: true },
  )
}

/** 新增网口（仅管理员） */
export function addVMInterface(name: string, data: VMInterfacePayload) {
  return service.post<unknown, ApiResponse<VMInterfaceInfo>>(
    `/vm/${encodeURIComponent(name)}/interfaces`,
    data,
  )
}

/** 更新网口（仅管理员） */
export function updateVMInterface(name: string, order: number, data: VMInterfacePayload) {
  return service.put<unknown, ApiResponse<null>>(
    `/vm/${encodeURIComponent(name)}/interfaces/${order}`,
    data,
  )
}

/** 删除网口（仅管理员） */
export function removeVMInterface(name: string, order: number) {
  return service.delete<unknown, ApiResponse<null>>(
    `/vm/${encodeURIComponent(name)}/interfaces/${order}`,
  )
}

File: server/model/vpc.go
package model

import "time"

// VPCSwitch 用户逻辑交换机配置。
type VPCSwitch struct {
	ID                         uint      `json:"id" gorm:"primaryKey"`
	Username                   string    `json:"username" gorm:"index;not null;size:64"`
	IsSystem                   bool      `json:"is_system" gorm:"default:false"`
	Name                       string    `json:"name" gorm:"index;not null;size:64"`
	BridgeName                 string    `json:"bridge_name" gorm:"not null;default:br-ovs;size:64"`
	BridgeMode                 string    `json:"bridge_mode" gorm:"not null;default:nat;size:16"`
	DHCPEnabled                bool      `json:"dhcp_enabled" gorm:"default:false"`
	UplinkMode                 string    `json:"uplink_mode" gorm:"not null;default:none;size:16"`
	UplinkIF                   string    `json:"uplink_if" gorm:"size:64"`
	UplinkGateway              string    `json:"uplink_gateway" gorm:"size:64"`
	OwnsBridge                 bool      `json:"owns_bridge" gorm:"default:false"`
	MigrateHostIP              bool      `json:"migrate_host_ip" gorm:"default:false"`
	HostAddrs                  string    `json:"-" gorm:"size:512"`
	HostGateway                string    `json:"-" gorm:"size:64"`
	HostMetric                 string    `json:"-" gorm:"size:16"`
	HostDNS                    string    `json:"-" gorm:"size:512"`
	LegacyMigrationRequired    bool      `json:"legacy_migration_required" gorm:"default:false"`
	BridgeVLANID               int       `json:"bridge_vlan_id" gorm:"default:0"`
	AllowPromiscuous           bool      `json:"allow_promiscuous" gorm:"default:false"`
	AllowMACChange             bool      `json:"allow_mac_change" gorm:"default:false"`
	AllowForgedTransmits       bool      `json:"allow_forged_transmits" gorm:"default:false"`
	IPv6SecurityEnabled        bool      `json:"ipv6_security_enabled" gorm:"default:false"`
	TrustedIPv6Prefixes        string    `json:"trusted_ipv6_prefixes" gorm:"type:text"`
	VLANID                     int       `json:"vlan_id" gorm:"uniqueIndex;not null"`
	CIDR                       string    `json:"cidr" gorm:"column:cidr;size:32"`
	GatewayIP                  string    `json:"gateway_ip" gorm:"not null;size:45"`
	DHCPStart                  string    `json:"dhcp_start" gorm:"not null;size:45"`
	DHCPEnd                    string    `json:"dhcp_end" gorm:"not null;size:45"`
	TrafficDownGB              float64   `json:"traffic_down_gb" gorm:"default:0"`
	TrafficUpGB                float64   `json:"traffic_up_gb" gorm:"default:0"`
	BandwidthMbps              int       `json:"bandwidth_mbps" gorm:"default:0"` // 兼容旧版单总带宽字段
	BandwidthDownMbps          int       `json:"bandwidth_down_mbps" gorm:"default:0"`
	BandwidthUpMbps            int       `json:"bandwidth_up_mbps" gorm:"default:0"`
	UsedTrafficDown            int64     `json:"used_traffic_down" gorm:"-"`
	UsedTrafficUp              int64     `json:"used_traffic_up" gorm:"-"`
	UsedTrafficDownGB          string    `json:"used_traffic_down_gb" gorm:"-"`
	UsedTrafficUpGB            string    `json:"used_traffic_up_gb" gorm:"-"`
	IsLimitedDown              bool      `json:"is_limited_down" gorm:"-"`
	IsLimitedUp                bool      `json:"is_limited_up" gorm:"-"`
	EffectiveBandwidthDownMbps int       `json:"effective_bandwidth_down_mbps" gorm:"-"`
	EffectiveBandwidthUpMbps   int       `json:"effective_bandwidth_up_mbps" gorm:"-"`
	CreatedAt                  time.Time `json:"created_at"`
	UpdatedAt                  time.Time `json:"updated_at"`
}

func (VPCSwitch) TableName() string {
	return "vpc_switches"
}

// VPCSecurityGroup 用户安全组。
type VPCSecurityGroup struct {
	ID         uint                   `json:"id" gorm:"primaryKey"`
	Username   string                 `json:"username" gorm:"index;not null;size:64"`
	VMName     string                 `json:"vm_name" gorm:"index;size:128"`
	Name       string                 `json:"name" gorm:"not null;size:64"`
	IsDefault  bool                   `json:"is_default" gorm:"default:false"`
	IsVMScoped bool                   `json:"is_vm_scoped" gorm:"default:false"`
	Remark     string                 `json:"remark" gorm:"size:255"`
	Rules      []VPCSecurityGroupRule `json:"rules,omitempty" gorm:"foreignKey:SecurityGroupID"`
	CreatedAt  time.Time              `json:"created_at"`
	UpdatedAt  time.Time              `json:"updated_at"`
}

func (VPCSecurityGroup) TableName() string {
	return "vpc_security_groups"
}

// VPCSecurityGroupRule 安全组规则。入站规则表示接收，出站规则表示拒绝；未命中规则时默认拒绝入站、允许出站。
type VPCSecurityGroupRule struct {
	ID              uint      `json:"id" gorm:"primaryKey"`
	SecurityGroupID uint      `json:"security_group_id" gorm:"index;not null"`
	Direction       string    `json:"direction" gorm:"size:16;not null"`                  // ingress/egress
	AddressFamily   string    `json:"address_family" gorm:"size:8;not null;default:ipv4"` // ipv4/ipv6
	Protocol        string    `json:"protocol" gorm:"size:16;not null"`                   // tcp/udp/icmp/icmpv6/all
	PortStart       int       `json:"port_start" gorm:"default:0"`
	PortEnd         int       `json:"port_end" gorm:"default:0"`
	TargetType      string    `json:"target_type" gorm:"size:32;not null"` // cidr/switch/security_group
	TargetValue     string    `json:"target_value" gorm:"size:128;not null"`
	Remark          string    `json:"remark" gorm:"size:255"`
	CreatedAt       time.Time `json:"created_at"`
	UpdatedAt       time.Time `json:"updated_at"`
}

func (VPCSecurityGroupRule) TableName() string {
	return "vpc_security_group_rules"
}

// VPCVMBinding VM 与交换机、安全组的绑定。每个 VM 可以有多个绑定（多网口），通过 interface_order 区分。
type VPCVMBinding struct {
	ID                   uint      `json:"id" gorm:"primaryKey"`
	VMName               string    `json:"vm_name" gorm:"uniqueIndex:idx_vm_interface;not null;size:128"`
	Username             string    `json:"username" gorm:"index;not null;size:64"`
	SwitchID             uint      `json:"switch_id" gorm:"index;not null"`
	SecurityGroupID      uint      `json:"security_group_id" gorm:"index;not null"`
	InterfaceOrder       int       `json:"interface_order" gorm:"uniqueIndex:idx_vm_interface;not null;default:0"`
	NicModel             string    `json:"nic_model" gorm:"size:32;default:virtio"`
	BandwidthInboundAvg  int       `json:"bandwidth_inbound_avg" gorm:"default:0"`
	BandwidthOutboundAvg int       `json:"bandwidth_outbound_avg" gorm:"default:0"`
	AllowedIPv4Addresses string    `json:"allowed_ipv4_addresses" gorm:"type:text"`
	AllowedIPv6Addresses string    `json:"allowed_ipv6_addresses" gorm:"type:text"`
	CreatedAt            time.Time `json:"created_at"`
	UpdatedAt            time.Time `json:"updated_at"`
}

func (VPCVMBinding) TableName() string {
	return "vpc_vm_bindings"
}

// VPCSwitchTrafficMonthly 交换机月流量记录。
type VPCSwitchTrafficMonthly struct {
	ID            uint      `json:"id" gorm:"primaryKey"`
	SwitchID      uint      `json:"switch_id" gorm:"uniqueIndex:idx_vpc_switch_month;not null"`
	Username      string    `json:"username" gorm:"index;not null;size:64"`
	Month         string    `json:"month" gorm:"uniqueIndex:idx_vpc_switch_month;not null;size:7"` // YYYY-MM
	TrafficDown   int64     `json:"traffic_down"`
	TrafficUp     int64     `json:"traffic_up"`
	OffsetDown    int64     `json:"offset_down"`
	OffsetUp      int64     `json:"offset_up"`
	IsLimitedDown bool      `json:"is_limited_down"`
	IsLimitedUp   bool      `json:"is_limited_up"`
	UpdatedAt     time.Time `json:"updated_at"`
}

func (VPCSwitchTrafficMonthly) TableName() string {
	return "vpc_switch_traffic_monthlies"
}
